### ISSUE TITLE
softmax_cross_entropy_with_logits to take integer labels / avoid giant dense one-hot matrix

### ISSUE THREAD

**Issue by mjwillson** [tensorflow/tensorflow#272]
Currently to compute this loss for a classifier you need to generate a big dense matrix of one-hot vectors to pass to softmax_cross_entropy_with_logits, as in the example at http://tensorflow.org/tutorials/mnist/tf/index.html#loss

This is kinda fiddly to do, but also rather wasteful of memory especially in the case of a larger multiclass softmax. Could we have a version of this function which accepts a list of indices, e.g. like Theano's: http://deeplearning.net/software/theano/library/tensor/nnet/nnet.html#tensor.nnet.categorical_crossentropy ?

**Comment by vrv** [comment]
This is a good suggestion -- (we named the existing op _with_logits because we anticipated this request, not because we like long function names ;).  I don't have a roadmap for exactly when this will be done though, but thanks for filing the feature request!

**Comment by mjwillson** [comment]
Cool, thanks!

Slightly confused by the logits thing -- in the docs that refers to the first argument (the input to the softmax) not the 'labels'/one-hot-vectors argument, no? Even then, the inputs to the softmax are unnormalized log-probabilities, but are they logits? I thought logits were log(p / (1-p)), rather than log p + const.

**Comment by chuanwen** [comment]
Is it possible to create a generic "select" function for tensor? Example usage:
    def loss(logits, labels):
         batch_size = tf.size(labels)
         logits_selected = tf.select(logits, tf.range(0, batch_size), labels)
         ....

**Comment by martinwicke** [comment]
@chuanwen: I'm not sure what exactly you mean, is that some combination of tf.where and tf.gather?

**Comment by vincentvanhoucke** [comment]
`sparse_softmax_cross_entropy_with_logits()` is now available
